### DISCUSSION TITLE
Location simulation

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#571] [Q&A] [answered]
Anyway much better for such quick tests is what Locus offers. Locus (tap the gps icon & turn off getting location) briefly disconnects the app from the gps data source. Your map center (North up & center cross shown) then becomes your ''virtual'' gps position.  By shifting the map position, you simulate to follow (deviate) a navigation track (with instructions) and quickly test any offtrack situation and the resulting (automatic) recalculation that way.
See the next video demo: https://youtu.be/r32a3VExhgg
An absolutely recommended and simple to use very usefull test tool.  Top !
Providing this similarly in the Cruiser app would definitely be very useful too.

_Originally posted by @0709wiwiwi in https://github.com/devemux86/cruiser/discussions/490#discussioncomment-9577299._

**Comment by devemux86** [comment]
- Long press on the location icon to select the location provider: GPS or Simulation

- In simulation the location is set at the map center while moving the map around

- The simulated location has a different color

- Available in location / advanced settings in https://github.com/devemux86/cruiser/discussions/559

<img src="https://github.com/devemux86/cruiser/assets/3484020/f47f35d2-4e8f-4e80-996b-5e738c6688cc" width="33%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/3f44a42d-c13a-4392-934c-3be799ea2f98" width="33%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/49c20d20-3552-4142-bc0b-c9f1b9d8c286" width="33%"/>

**Reply by Ciutnik** [reply]
Interesting. What speed does the simulation assume? If it could be variable, we could also test the autozoom behaviour 🙂

**Reply by devemux86** [reply]
Based on scroll distance and time, but if we move the map at small zooms quickly, it may calculate airplane speed! :slightly_smiling_face: 

At large zooms it should be more real.

**Reply by Ciutnik** [reply]
Great! So full simulation will be possible 👍

The "Mock Location" simulator I use in Android, in addition to being able to "drive" along a loaded GPX track (it can also route itself), slows down before sharp corners, giving the simulation realism. This is how I have tested autozoom so far 😃

**Comment by devemux86** [comment]
Location simulation is available for testing in:

- https://github.com/devemux86/cruiser/discussions/610

- Instructions: https://github.com/devemux86/cruiser/discussions/571#discussioncomment-9787299

**Comment by devemux86** [comment]
Has anyone tried this feature and find it useful?

I'm thinking if we really need it or it's better to remove it and not confuse users.

**Reply by Ciutnik** [reply]
I admit that I completely forgot about this new feature, but I also haven't used the simulation for a long time. I'm sure it will change with each passing day, as the weather gets worse...

Unfortunately, as it is now, although it works great, it does not allow you to check the zoom or rotation of the map during navigation. This is probably the hardest to imagine and predict. Perhaps a "virtual gas pedal" would be more useful when simulating navigation 😉

**Reply by devemux86** [reply, marked as answer]
Indeed it is simple, so we better think the implementation before making it public,
as other features have more priority.

---

For development, I use a different simulator that I created before the navigator.
It is very advanced to simulate everything in real life conditions in whole world.

And this can also work with external controllers, keyboards, driving wheels, etc. :slightly_smiling_face:

**Reply by Ciutnik** [reply]
> And this can also work with external controllers, keyboards, driving wheels, etc.

It's like playing on the machines in old-time arcades. It's almost virtual reality! I imagine a winter evening in front of the TV screen, and in it the Cruiser is running with the steering wheel connected. Hot tea, good company, and there is no hell, so everything is permitted 😉
Just don't get into bad habits before the new season in real reality 🙂
